SPEC.md

```diff
@@ -0,0 +1,176 @@
+    - **Network namespace path**. This represents the path to the network namespace to be added, i.e. /proc/[pid]/ns/net or a bind-mount/link to it.
+    - **Network configuration**. This is a JSON document describing a network to which a container can be joined. The schema is described below.
+    - **Extra arguments**. This allows granular configuration of CNI plugins on a per-container basis.
+    - **Name of the interface inside the container**. This is the name that should be assigned to the interface created inside the container (network namespace); consequently it must comply with the standard Linux restrictions on interface names.
+  - Result:
+    - **IPs assigned to the interface**. This is either an IPv4 address, an IPv6 address, or both.
+
+- Delete container from network
+  - Parameters:
+    - **Container ID**, as defined above.
+    - **Network namespace path**, as defined above.
+    - **Network configuration**, as defined above.
+    - **Extra arguments**, as defined above.
+    - **Name of the interface inside the container**, as defined above.
+
+The executable command-line API uses the type of network (see [Network Configuration](#network-configuration) below) as the name of the executable to invoke. It will then look for this executable in a list of predefined directories. Once found, it will invoke the executable using the following environment variables for argument passing:
+- `CNI_COMMAND`: indicates the desired operation; either `ADD` or `DEL`
+- `CNI_CONTAINERID`: Container ID
+- `CNI_NETNS`: Path to network namespace file
+- `CNI_IFNAME`: Interface name to set up
+- `CNI_ARGS`: Extra arguments passed in by the user at invocation time. Alphanumeric key-value pairs separated by semicolons; for example, "FOO=BAR;ABC=123"
+- `CNI_PATH`: Colon-separated list of paths to search for CNI plugin executables
+
+Network configuration in JSON format is streamed through stdin.
+
+Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command:
+```json
+{
+    "ip4": <ipv4-and-subnet-in-CIDR>,
+    "ip6": <ipv6-and-subnet-in-CIDR>
+}
+```
+
+Errors are indicated by a non-zero return code and error details printed to stderr.
+
+### Network Configuration
+
+The network configuration is described in JSON form. The configuration can be stored on disk or generated from other sources by the container runtime. The following fields are well-known and have the following meaning:
+- `name` (string): Network name. This should be unique across all containers on the host (or other administrative domain).
+- `type` (string): Refers to the filename of the CNI plugin executable.
+- `ipMasq` (boolean): Optional (if supported by the plugin). Set up an IP masquerade on the host for this network.
```

Comment:
The reason we need to support IP masquerade is because very often the IPs assigned to the interface are not routable outside their subnet. e.g., that's an issue with Docker, no news here. Even with something like flannel, these IPs are not routable outside of flannel network. So if a network/user wishes to provide access to "outside" subnets, the container's IP needs to be NAT'ed at the host. That's what this flag is for.

Now, this flag is completely up to the plugin to implement or not. In general, this spec does not want to define fields that are plugin specific. However, to avoid every plugin calling it differently, we decided to standardize on the name.